Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp

```diff
@@ -368,15 +360,144 @@ namespace AZ::DocumentPropertyEditor
         }
     }
 
-    void RowFilterAdapter::UpdateMatchState(MatchInfoNode* rowState)
+    void RowFilterAdapter::GenerateFullTree()
     {
-        auto includeInFilter = [](MatchInfoNode* node) -> bool
+        AZ_Assert(!m_root, "GenerateFullTree must only be called on an empty tree!");
+        m_root = MakeNewNode(nullptr, m_updateFrame);
+        const auto& sourceContents = m_sourceAdapter->GetContents();
+
+        // we can assume all direct children of an adapter must be rows; populate each of them
+        for (size_t topIndex = 0; topIndex < sourceContents.ArraySize(); ++topIndex)
+        {
+            PopulateNodesAtPath(Dom::Path({ Dom::PathEntry(topIndex) }), false);
+        }
+    }
+
+    void RowFilterAdapter::CullUnmatchedChildRows(Dom::Value& rowValue, const MatchInfoNode* rowMatchNode)
+    {
+        auto& matchChildren = rowMatchNode->m_childMatchState;
+        const bool sizesMatch = (rowValue.ArraySize() == matchChildren.size());
+        AZ_Assert(sizesMatch, "DOM value and cached match node should have the same number of children!");
+
+        if (sizesMatch)
+        {
+            auto valueIter = rowValue.MutableArrayBegin();
+            for (auto matchIter = matchChildren.begin(); matchIter != matchChildren.end(); ++matchIter)
+            {
+                // non-row children will have a null MatchInfo -- we only need to worry about culling row values
+                auto currMatch = *matchIter;
+                if (currMatch != nullptr)
+                {
+                    if (currMatch->IncludeInFilter())
+                    {
+                        CullUnmatchedChildRows(*valueIter, *matchIter);
+                    }
+                    else
+                    {
+                        valueIter = rowValue.ArrayErase(valueIter);
+                        continue; // we've already moved valueIter forward, skip the ++valueIter below
+                    }
+                }
+                ++valueIter;
+            }
+        }
+    }
```

Comment:
This function doesn't touch any member variables.
It can be marked as static.

Reply:
good call, fixed!